RHIDP-2024: Configuring PostgreSQL databases in RHDH (#262)
redhat-developer/red-hat-developers-documentation-rhdh: modules/admin/proc-configuring-postgresql-instance-using-operator.adoc

[id="proc-configuring-postgresql-instance-using-operator_{context}"]
= Configuring an external PostgreSQL instance using the Operator

You can configure an external PostgreSQL instance by using the {product} Operator. By default, the Operator creates and manages a local instance of PostgreSQL in the same namespace where you have deployed the {product-very-short} instance. However, you can change this default setting to configure an external PostgreSQL database server, for example, AWS RDS or Azure database. 

.Prerequisites

* You are using a supported version of PostgreSQL. For more information, see the link:https://access.redhat.com/support/policy/updates/developerhub[Product life cycle page].
* You have the following details:
** `db-host`: Denotes your PostgreSQL instance Domain Name System (DNS) or IP address 
** `db-port`: Denotes your PostgreSQL instance port number, usually `5432`
** `username`: Denotes the user name to connect to your PostgreSQL instance
** `password`: Denotes the password to connect to your PostgreSQL instance

> * You have installed the {product-very-short} application by using the Operator or Helm Chart.

Reviewer: The prerequisite shouldn't be to install the RHDH instance, because that would lead to a conflict. You want to have the operator installed, but not the RHDH itself.

```suggestion
* You have installed the Operator for {product-very-short}.
```

  <certificate-key> <2>
  # ... 
EOF  
----
<1> Provide the name of the certificate secret.
<2> Provide the certificate key.

. Create a credential secret to connect with the PostgreSQL instance:
+
[source,yaml]
----
cat <<EOF | oc -n <your-namespace> create -f -
apiVersion: v1
kind: Secret
metadata:
 name: <cred-secret> <1>
type: Opaque
stringData: <2>
 POSTGRES_PASSWORD: <password>
 POSTGRES_PORT: "<db-port>"
 POSTGRES_USER: <username>
 POSTGRES_HOST: <db-host>
 PGSSLMODE: <ssl-mode> # for TLS connection <3>
 NODE_EXTRA_CA_CERTS: <abs-path-to-pem-file> # for TLS connection, e.g. /opt/app-root/src/postgres-crt.pem <4>
EOF
----
<1> Provide the name of the credential secret.
<2> Provide credential data to connect with your PostgreSQL instance.
<3> Optional: Provide the value based on the required link:https://www.postgresql.org/docs/15/libpq-connect.html#LIBPQ-CONNECT-SSLMODE[SSL mode].
<4> Optional: Provide the value only if you need a TLS connection for your PostgreSQL instance.

. Create a `Backstage` custom resource (CR):
+
[source,yaml]
----
cat <<EOF | oc -n <your-namespace> create -f -
apiVersion: rhdh.redhat.com/v1alpha1
kind: Backstage
metadata:

>  name: <backstage-instance-name>

Reviewer: Fix YAML indentation by increments of 2 spaces (`  `).

[id="proc-configuring-postgresql-instance-using-operator_{context}"]
= Configuring an external PostgreSQL instance using the Operator

> You can configure an external PostgreSQL instance by using the {product} Operator. By default, the Operator creates and manages a local instance of PostgreSQL in the same namespace where you have deployed the {product-very-short} instance. However, you can change this default setting to configure an external PostgreSQL database server, for example, Amazon Web Services (AWS) Relational Database Service (RDS) or Azure database.

Reviewer 2: ```suggestion
You can configure an external PostgreSQL instance using the {product} Operator. By default, the Operator creates and manages a local instance of PostgreSQL in the same namespace where you have deployed the {product-very-short} instance. However, you can change this default setting to configure an external PostgreSQL database server, for example, Amazon Web Services (AWS) Relational Database Service (RDS) or Azure database.
```

.Prerequisites

* You are using a supported version of PostgreSQL. For more information, see the link:https://access.redhat.com/support/policy/updates/developerhub[Product life cycle page].
* You have the following details:
** `db-host`: Denotes your PostgreSQL instance Domain Name System (DNS) or IP address 
** `db-port`: Denotes your PostgreSQL instance port number, usually `5432`
** `username`: Denotes the user name to connect to your PostgreSQL instance
** `password`: Denotes the password to connect to your PostgreSQL instance
* You have installed the {product} Operator.
* Optional: You have a CA certificate, TLS private key, and TLS certificate so that you can secure your database connection by using the Transport Layer Security (TLS) protocol. For more information, refer to your PostgreSQL vendor documentation.

[NOTE]
====

> By default, {product-short} uses a database for each plugin and automatically creates it if none is found. As an administrator, know that users might need the `Create Database` privilege in addition to `PSQL Database` privileges for configuring an external PostgreSQL instance.  

Reviewer 2: ```suggestion
By default, {product-short} uses a database for each plugin and automatically creates it if none is found. You might need the `Create Database` privilege in addition to `PSQL Database` privileges for configuring an external PostgreSQL instance.  
```

====
 

.Procedure

> . Optional: Create a certificate secret to configure your PostgreSQL instance with a TLS connection:

Reviewer 2: I'd suggest adding the full form of TLS in the first instance.

Author: expanded at the first instance in the module.

----
cat <<EOF | oc -n <your-namespace> create -f -
apiVersion: v1
kind: Secret
metadata:
 name: <crt-secret> <1>
type: Opaque
stringData:
 postgres-ca.pem: |-
  -----BEGIN CERTIFICATE-----
  <ca-certificate-key> <2>
 postgres-key.key: |-
  -----BEGIN CERTIFICATE-----
  <tls-private-key> <3>
 postgres-crt.pem: |-    
  -----BEGIN CERTIFICATE-----
  <tls-certificate-key> <4>
  # ...
EOF
----
<1> Provide the name of the certificate secret.
<2> Provide the CA certificate key.
<3> Optional: Provide the TLS private key.
<4> Optional: Provide the TLS certificate key.

. Create a credential secret to connect with the PostgreSQL instance:
+
[source,yaml]
----
cat <<EOF | oc -n <your-namespace> create -f -
apiVersion: v1
kind: Secret
metadata:
 name: <cred-secret> <1>
type: Opaque
stringData: <2>
 POSTGRES_PASSWORD: <password>
 POSTGRES_PORT: "<db-port>"
 POSTGRES_USER: <username>
 POSTGRES_HOST: <db-host>
 PGSSLMODE: <ssl-mode> # for TLS connection <3>
 NODE_EXTRA_CA_CERTS: <abs-path-to-pem-file> # for TLS connection, e.g. /opt/app-root/src/postgres-crt.pem <4>
EOF
----
<1> Provide the name of the credential secret.
<2> Provide credential data to connect with your PostgreSQL instance.

> <3> Optional: Provide the value based on the required link:https://www.postgresql.org/docs/15/libpq-connect.html#LIBPQ-CONNECT-SSLMODE[SSL mode].

Reviewer 2: Please see the previous suggestion about adding the full form of SSL.

<4> Optional: Provide the value only if you need a TLS connection for your PostgreSQL instance.

> . Create a `Backstage` custom resource (CR):

Reviewer 2: It is really Backstage CR? or RHDH CR?

Reviewer 3: ![image](https://github.com/redhat-developer/red-hat-developers-documentation-rhdh/assets/227597/665bcf01-3fdb-489d-8990-3afe7b9b6557)

+
[source,yaml]
----
cat <<EOF | oc -n <your-namespace> create -f -
apiVersion: rhdh.redhat.com/v1alpha1
kind: Backstage
metadata:
  name: <backstage-instance-name>
spec:
  database:
    enableLocalDb: false <1>
  application: 
    extraFiles:
      mountPath: <path> # e g /opt/app-root/src
      secrets:
        - name: <crt-secret> <2>
          key: postgres-crt.pem, postgres-ca.pem, postgres-key.key # key name as in <crt-secret> Secret
    extraEnvs:
      secrets:
        - name: <cred-secret> <3>
        # ... 
----
<1> Set the value of the `enableLocalDb` parameter to `false` to disable creating local PostgreSQL instances.
<2> Provide the name of the certificate secret if you have configured a TLS connection. 
<3> Provide the name of the credential secret that you created.
+
[NOTE]
====

> The environment variables listed in the `Backstage` CR work with the Operator default configuration. If you have changed the Operator default configuration, you must reconfigure the `Backstage` CR accordingly.

Reviewer 2: Please confirm with SME if this is really Backstage CR or Developer Hub CR?

Author: @zdrapela, Should we mention  `Developer Hub` CR rather than using `Backstage` CR here? Please confirm.

Author: I think we should use the exact CR name. the link mentioned here also point to the `Backstage` CR, not the `Developer Hub` CR. Throughout the doc, we have used the name `Backstage` CR, not `Developer Hub` CR in any code snippet.

Reviewer: @kim-tsao ^^?

Reviewer 3: hey folks, to confirm what you see in the UI, you can:
* launch a cluster (eg., send a message to clusterbot on slack `launch 4.16 gcp`
* wait until the cluster spins up
* install the operator subscription from https://github.com/janus-idp/operator/blob/main/.rhdh/docs/installing-ci-builds.adoc
* once the suubscription is created, switch to the Developer perspective, open your chosen namespace, hit the `Add+` button, and scroll down to `Operator-Backed`
* search for developer hub, and when creating/configuring it, note what it says in the YAML view:

I believe you'll see this - even thouigh it's a custom resource for RHDH, the `KIND` is still `Backstage`:

```
kind: Backstage
```

Author: The YAML view shows the `Kind` field set to `Backstage` resource.
[
![Backstage-Cr](https://github.com/redhat-developer/red-hat-developers-documentation-rhdh/assets/94683525/afa033d9-5224-4f75-9931-9e6202e1bd6a)

Reviewer 3: ![image](https://github.com/redhat-developer/red-hat-developers-documentation-rhdh/assets/227597/3415f1df-07cb-4381-8300-2265756414b2)

.Prerequisites

> * You are using PostgreSQL version 15.

Reviewer 4: Per previous comment, Postgres 14, 15, 16 are supported.  We should just state a supported version of PostgreSQL.
